spec/abci++/abci++_app_requirements_002_draft.md

```diff
@@ -177,3 +179,818 @@ Likewise, `ExtendVote` can also be non-deterministic:
+not broadcasted to other peers and not included in a proposal block.
+
+`Data` contains the result of the `CheckTx` transaction execution, if any. It is not required to be
+deterministic since, given a transaction, nodes' Applications
+might have a different `CheckTxState` when they receive it and check their validity via `CheckTx`.
+It is
+semantically meaningless to Tendermint.
+
+`Events` include any events for the execution, though since the transaction has not
+been committed yet, they are effectively ignored by Tendermint.
+
+In v0.35.x and later, there is a `Priority` field in `ResponseCheckTx` that can be
+used to explicitly prioritize transactions in the mempool for inclusion in a block
+proposal.
+
+#### Specifics of `ExecTxResult`
+
+`FinalizeBlock` is the workhorse of the blockchain. Tendermint delivers the decided block,
+including the list of all its transactions synchronously to the Application.
+The block delivered (and thus the transaction order) is the same at all correct nodes as guaranteed
+by the Agreement property of Tendermint consensus.
+
+In same block execution mode `ExecTxResult` is returned by `PrepareProposal` so that its data can
+be used to build the block to be proposed in the current height.
+
+The `Data` field in `ExecTxResult` contains an array of bytes with the transaction result.
+It must be strictly deterministic (all nodes should return the same value), but can contain arbitrary
+data. Likewise, the value of `Code` must be deterministic.
+Both the `Code` and `Data` are included in a structure that is hashed into the
+`LastResultsHash` of the block header in the next height (next block execution mode), or the
+header of the block to propose in the current height (same block execution mode, `ExecTxResult` as
+part of `PrepareProposal`).
+
+If `ExecTxResult` contains `Code != 0`, the transaction will be considered invalid,
+though it is still included in the block.
+
+`Events` include any events for the execution, which Tendermint will use to index
+the transaction by. This allows transactions to be queried according to what
+events took place during their execution.
+
+### Updating the Validator Set
+
+The application may set the validator set during
+[`InitChain`](./abci%2B%2B_methods_002_draft.md#initchain), and may update it during
+[`FinalizeBlock`](./abci%2B%2B_methods_002_draft.md#finalizeblock)
+(next block execution mode) or
+[`PrepareProposal`](./abci%2B%2B_methods_002_draft.md#prepareproposal)/[`ProcessProposal`](./abci%2B%2B_methods_002_draft.md#processproposal)
+(same block execution mode). In all cases, a structure of type
+[`ValidatorUpdate`](./abci%2B%2B_methods_002_draft.md#validatorupdate) is returned.
+
+The `InitChain` method can return a list of validators.
+If the list is empty, Tendermint will use the validators loaded in the genesis
+file.
+If the list returned by `InitChain` is not empty, Tendermint will use its contents as the validator set.
```

Comment:
Question: is it a list of validators or validator updates? Is the list *adopted* by Tendermint, or are the operations (updates) performed/implemented? What happens if this list is not well formed?

Reply:
That's a good question. I'll answer it here, but think getting into the details of this here would be overkill (it should be explained in the "methods" section (for `InitChain`).

The answer: you're right in that it's updates. In normal execution that update contains (a) new validators and their power (the validator MUST NOT exist in the set before the update), (b) existing validators with modified power (power is overwritten on top of the previous one), (c) validators with power set to 0 (those validators MUST exist in the set before update; they are removed from the set).

In this case, we use the same semantics, with the empty set as the set before the update, and the update only containing (a) updates.

So when you look at the update struct (with only (a) updates) it's pretty similar to the set it updates to.

Reply:
Actually, let me add something like this in the "Methods" section in this PR, even if it's not part of ABCI++ (strictly speaking)